### PR TITLE
Update for new Ruby and Rails versions

### LINKED ISSUE #498: File.exists? is deprecated
`File.exists?` is deprecated, and this prints out a deprecation warning on Ruby 3.0+. Change to `File.exist?` everywhere.

### PR DESCRIPTION
**Pull request by BrianHawley**
- Test Ruby 2.7, 3.0, and 3.1, and Rails 6.1 and 7.0.
- Exclude incompatible test matrix combinations.
- Show Ruby deprecation warnings in tests where supported.
- File.exists? is deprecated; use File.exist?.
- URI.escape and URI.decode are deprecated.
- Add frozen string literal comments.
- Make sure constants are frozen.
- Make sure that the code is frozen-string safe.
- Adjust dev dependencies for Ruby 3.0+ compatibility.
- Fix Ruby warnings in pdfkit_spec.rb
- Comment out bad syntax in release-drafter action.

Fixes #498

### CODE REVIEW

**Comment by BrianHawley** [comment]
@serene I ran all the tests locally in all version combinations so I could make sure the deprecations were fixed. Once Travis runs, we should be able to see whether the config changes worked.

**Comment by BrianHawley** [comment]
Since Ruby 3.1 is out now, I updated the PR to include it in its tests. The tests pass locally with no warning.

**Comment by serene** [comment]
I will add this to my list to review, but if others can put some eyes on it as well, I'd really appreciate it!

File: lib/pdfkit/source.rb [hunk 1]
```diff
@@ -38,11 +42,11 @@ def to_s
     private
 
     def shell_safe_url
-      url_needs_escaping? ? URI::escape(@source) : @source
+      url_needs_escaping? ? URI::DEFAULT_PARSER.escape(@source) : @source
```
Comment (BrianHawley) [inline lib/pdfkit/source.rb:45]: This is a direct fix of the `URI.escape` deprecation, and has the same exact behavior, but my main concern is that I don't think this is the right escaping method.

`URI.escape` escapes a string for use in a URI as a value. However, in this case, the string is being escaped for use in a shell command parameter, which doesn't really have the same requirements. For one thing, the percent-encoding that `URI.escape` generates might conflict with the `%var` environment variable syntax on Windows. There might be some other issues, but I haven't taken the time to check.

The change here has no effect on the existing behavior (except getting rid of the deprecation message), so it's probably the best to do here. However, in the long run, it's probably best to change to a more appropriate escaping method.

**Comment by BrianHawley** [comment]
I notice that the Travis CI tests didn't run. Should we set up Github Action tests?

**Comment by serene** [comment]
@BrianHawley if that's something you'd like to take on, that would be really great! Maybe as a separate PR?

**Comment by BrianHawley** [comment]
@serene yes, preferably as a separate PR. For one thing, I haven't done that before, so I'd have to take time to learn how :)

**Comment by BrianHawley** [comment]
I've been using a patched version of pdfkit with these changes for a month (used vendorer to automatically generate the patched version). Are we going to merge this soon and release a new version, so I can switch back to the regular gem?

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -6,13 +8,15 @@
   add_filter 'spec/'
 end
 
+Warning[:deprecated] = true if defined?(Warning.[]=)
```
Comment (thomasdarde) [inline spec/spec_helper.rb:11]: Nice PR @BrianHawley 
What is the use of this line about deprecated ?

**Comment by thomasdarde** [comment]
Seems really nice, I did also use this version without any issues !

File: spec/spec_helper.rb (hunk 2 above)
Comment (BrianHawley) [reply spec/spec_helper.rb:11]: That makes Ruby deprecation warnings show up while running tests. I used those warnings to determine and fix any future compatibility issues, when I was running the tests locally against all supported versions. It's a good trick I use at work as well.

**Comment by dedene** [comment]
We are also using this version without any issues. Can't it be merged and released?

**Comment by leonyip** [comment]
@serene is it possible that this repo had not migrated to travis-ci.com? https://blog.travis-ci.com/2021-05-07-orgshutdown

**Comment by leonyip** [comment]
@serene I just opened #502 for the Github Actions migration. This branch can be tested against that and should be able to merge. Thanks!

**Comment by BrianHawley** [comment]
Pushed up a rebased version with the same test matrix changes applied to the Github Actions tests. The workflow needs approval before the tests will run.

**Comment by BrianHawley** [comment]
Cool, looks like the exclusions worked properly.

**Comment by BrianHawley** [comment]
OK, that's not flaky. Let me check if another exclusion is needed.

**Comment by serene** [comment]
Should we even be supporting 2.5 anymore?

**Comment by BrianHawley** [comment]
I don't have a problem with leaving 2.5 support there. I'd rather keep it there long enough to figure out what the problem was. That failure was uninformative.

**Comment by leonyip** [comment]
@BrianHawley I sent a PR to your repo with the exclusions. It ran fine on my build.

**Comment by leonyip** [comment]
Might have something to do with the bundler cache, can you omit the `bundler-cache: true`

**Comment by BrianHawley** [comment]
OK, so that didn't work. I'll try something else after the meeting I'm in now.

**Comment by BrianHawley** [comment]
@serene I had to change the Ruby version numbers to strings to make sure 3.0 tests 3.0, instead of 3 (which becomes 3.1). Could you approve the workflow again?

**Pull request merged**